core/primitives/src/types.rs

```diff
@@ -977,3 +977,13 @@ pub trait EpochInfoProvider {
 
     fn minimum_stake(&self, prev_block_hash: &CryptoHash) -> Result<Balance, EpochError>;
 }
+
+/// State of the trie cache.
+#[derive(Debug, Copy, Clone)]
+pub enum TrieCacheState {
+    /// We put each visited node to LRU cache. It generally saves time, but existence of any node is not guaranteed;
+    CachingShard,
+    /// We put each visited node to the chunk cache which is a hash map. On the chunk processing, we guarantee that all
+    /// nodes visited on such state will remain in the chunk cache.
```

Comment:
Probably the second sentence worth rephrasing. Like this:

This is needed to guarantee the all nodes are retrieved at most once during a single chunk processing. In other words we guarantee that any node visited during chunk processing lives in cache until the chunk processing is finished. The guarantee is needed so we can compute the gas in a way that user doesn't pay twice for a single node retrieval.

Reply:
That's not exactly true. We don't (currently can't) cache all nodes visited in a chunk. What we want to achieve is the implication "CachingChunk state is enabled => Each node is retrieved from DB at most once".

But yeah, I use it to improve readability.